<h1 dir="rtl">פרוייקט סיום בהשתלמות מורים בנושא למידה עמוקה</h1>

<div dir="rtl">
<table dir>
  <tr>
    <td>מגיש</td>
    <td>איתי בארלי</td>
  </tr>
  <tr>
    <td>מספר ת"ז</td>
    <td>032866204</td>
  </tr>
  <tr>
    <td>תאריך הגשה</td>
    <td>1 ספטמבר 2024</td>
  </tr>
  <tr>
    <td>מורים</td>
    <td>אריאל בר-יצחק, אורן עוזרד</td>
  </tr>
</table>
</div>

<h2 dir="rtl">הנתונים</h2>

<div dir="rtl">
<table dir>
  <tr>
    <td>נושא</td>
    <td>תחרויות שחמט</td>
  </tr>
  <tr>
    <td>אתר</td>
    <td>Kaggle</td>
  </tr>
  <tr>
    <td>כותרת הדף</td>
    <td>Chess Game Dataset</td>
  </tr>
  <tr>
    <td>תאריך עדכון אחרון</td>
    <td>19 ספטמבר 2020</td>
  </tr>
  <tr>
    <td>שם בעל הדף</td>
    <td>Praveen Kumar</td>
  </tr>
  <tr>
    <td>קישור</td>
    <td><a href="https://www.kaggle.com/datasets/penchalaiah123/chess-game-dataset/data">Kaggle</a></td>
  </tr>
  <tr>
    <td>קישור חלופי</td>
    <td><a href="https://raw.githubusercontent.com/ItaiBTeacher/ml-2324/main/projects/chess/data/static/dataset.csv">GitHub</a></td>
  </tr>
  <tr>
    <td>מאיפה נאספו הנתונים</td>
    <td>מהאתר <a href="lichess.org">lichess.org</a> המשמש למשחקי שח אונליין</td>
  </tr>
</table>
</div>

<h2 dir="rtl">ייבוא ספריות</h2>

In [125]:
import datetime as dt
import numpy as np # numerical computations
import pandas as pd # DataFrames and Series
import tensorflow as tf # Deep learning backend

# The following two lines were suggested to me
# by the 'Configuring your backend' section
# of the 'Getting Started with Keras' webpage
# on Keras' official website.
# https://keras.io/getting_started/
import os
os.environ["KERAS_BACKEND"] = "tensorflow"
import keras as kr # Deep learning frontend

<h2 dir="rtl">מבנה הטבלה</h2>

<p dir="rtl">טעינת הטבלה</p>

In [5]:
orig = pd.read_csv('https://raw.githubusercontent.com/ItaiBTeacher/ml-2324/main/projects/chess/data/static/dataset.csv')

<p dir="rtl">מספר שורות</p>

In [6]:
orig.shape[0]

20058

<p dir="rtl">מספר עמודות</p>

In [24]:
len(orig.columns)

16

<div dir="rtl">
    <p>שמות העמודות, טיפוסי הנתונים שלהן, ומספר הערכים החסרים (NaN) בכ"א מהן</p>
    <p>ניתן לראות שאין באף עמודה ערכים חסרים.</p>
</div>

In [12]:
orig.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              20058 non-null  object 
 1   rated           20058 non-null  bool   
 2   created_at      20058 non-null  float64
 3   last_move_at    20058 non-null  float64
 4   turns           20058 non-null  int64  
 5   victory_status  20058 non-null  object 
 6   winner          20058 non-null  object 
 7   increment_code  20058 non-null  object 
 8   white_id        20058 non-null  object 
 9   white_rating    20058 non-null  int64  
 10  black_id        20058 non-null  object 
 11  black_rating    20058 non-null  int64  
 12  moves           20058 non-null  object 
 13  opening_eco     20058 non-null  object 
 14  opening_name    20058 non-null  object 
 15  opening_ply     20058 non-null  int64  
dtypes: bool(1), float64(2), int64(4), object(9)
memory usage: 2.3+ MB


<p dir="rtl">
נעבור עמודה עמודה ונראה מה משמעותה ואילו ערכים היא מכילה. בסוף נסכם את כל תובנותינו בטבלה.
</p>

<h3 dir="trl">כל עמודה בנפרד</h3>

<h4 dir="trl">id</h4>

<h5 dir="trl">טיפוס הנתונים</h5>

In [25]:
orig.id.dtype

dtype('O')

<h5 dir="trl">מספר ערכים שונים</h5>

In [26]:
orig.id.nunique()

19113

<h5 dir="trl">ערכים לדוגמה</h5>

In [28]:
orig.id.unique()[:10]

array(['TZJHLljE', 'l1NXvwaE', 'mIICvQHh', 'kWKvrqYL', '9tXo1AUZ',
       'MsoDV9wj', 'qwU9rasv', 'RVN0N3VK', 'dwF3DJHO', 'afoMwnLg'],
      dtype=object)

<h5 dir="trl">משמעות העמודה</h5>

<div dir="rtl">
<p>
הנתונים בטבלה לקוחים מהאתר
lichess.org
שמהווה פלטפורמה למשחקי שחמט מקוונים.
כל שורה בטבלה מתארת משחק ששוחק באתר זה.
לכל משחק יש מזהה, שהוא מחרוזת, למשל
TZJHLljE, l1NXvwaE, mIICvQHh.
</p>
<p>
לאור זאת הייתי מצפה שלכל שתי שורות שונות בטבלה יהיו ערכים שונים בעמודה
id.
אך בפועל מספר הערכים השונים בעמודה זו
(19113)
קטן ממספר השורות בטבלה
(20053).
מכאן שיש מזהי משחק המופיעים פעמיים או יותר. נוודא זאת בקוד.
</p>
</div>

In [60]:
orig.id.value_counts().head()

id
XRuQPSzH    5
edYOVb5F    4
o4T7dj5h    4
I8odmGjb    4
ZPHBiKBY    4
Name: count, dtype: int64

<div dir="rtl">
<p>
אכן, יש מזהי משחק המופיעים יותר מפעם אחת, לדוגמה
XRuQPSzH
מופיע 5 פעמים.
</p>
<p>
מכל שורה שיש לה כפילויות בטבלה נשאיר רק עותק אחד. אולי זה יפתור את הבעיה.
</p>
</div>

In [68]:
orig_no_dup_rows = orig.drop_duplicates()

<p dir="rtl">
האם זה פתר את הבעיה?
</p>

In [69]:
print('Number of rows: ', orig_no_dup_rows.shape[0])
print('Number of unique ids: ', orig_no_dup_rows.id.nunique())

Number of rows:  19629
Number of unique ids:  19113


<p dir="rtl">
לא פתר את הבעיה! בכמה שורות סה"כ מופיעים מזהי משחק שאינם יחודיים בעמודה
id?
ומה החלק היחסי של שורות אלו ביחס למספר השורות הכולל בטבלה?
</p>

In [74]:
dup_id_row_num = sum(orig_no_dup_rows.id.duplicated(keep=False))
print('Number of rows with a non-unique id: ', dup_id_row_num)
print('Which is ', (dup_id_row_num / orig_no_dup_rows.shape[0]) * 100, ' percent of the total number of table rows')

Number of rows with a non-unique id:  1032
Which is  5.2575271282286415  percent of the total number of table rows


<p dir="rtl">
ננקוט בצעד הפשוט ונשאיר מכל מקבץ שורות שמזהה המשחק שלהן מופיע יותר מפעם אחת רק את השורה הראשונה במקבץ. בעשותנו כך, יימחקו מהטבלה לכל היותר 5.26% מכלל השורות בטבלה, וזה יותר אותנו עם למעלה מ- 18000 שורות.
</p>

In [75]:
orig_no_dup_ids = orig_no_dup_rows.drop_duplicates(subset='id')

<p dir="rtl">
עתה מספר השורות בטבלה זהה למספר מזהי המשחק השונים בעמודה
id.
</p>

In [76]:
print('Number of rows: ', orig_no_dup_ids.shape[0])
print('Number of unique ids: ', orig_no_dup_ids.id.nunique())

Number of rows:  19113
Number of unique ids:  19113


<p dir="rtl">
עתה בעמודה
id
אין מידע שעשוי לתרום ללמידה, ולפיכך נשמיט אותה מהטבלה.
</p>

In [77]:
orig_no_ids = orig_no_dup_ids.drop(labels='id', axis=1)

<h4 dir="trl">rated</h4>

<h5 dir="trl">טיפוס הנתונים</h5>

In [120]:
orig_no_ids.rated.dtype

dtype('bool')

<h5 dir="trl">מספר ערכים שונים</h5>

In [119]:
orig_no_ids.rated.nunique()

2

<h5 dir="trl">רשימת כל הערכים</h5>

In [118]:
orig_no_ids.rated.unique()

array([False,  True])

<h5 dir="trl">משמעות העמודה</h5>

<p dir="rtl">
לכל השחקנים בטבלה יש דירוג (ר' להלן בתיאור העמודה
white_rating).
הדירוג של שחקן עשוי להשתנות בתום כל משחק שהוא השתתף בו.
הדירוג של שחקן בתום משחק כלשהו הוא פונקציה של דירוג השחקן לפני תחילת המשחק
ושל מאפייני המשחק שזה עתה הסתיים
(למשל אם השחקן ניצח במשחק או הפסיד בו, ומה היה דירוג היריב שלו בתחילת המשחק).
יש משחקים שנקבע לגביהם מלכתחילה שלא ישפיעו על דירוג השחקנים.
משחקים אלו מצויינים בטבלה ע"י הערך
False
בעמודה
rated.
המשחקים האחרים, כלומר אלו שהוחלט לגביהם מלכתחילה שישוכללו בדירוג השחקנים המשתתפים בהם,
מצויינים בטבלה ע"י הערך
True
בעמודה
rated.
</p>

<h5 dir="rtl">סימוכין</h5>

<a href="https://www.chess.com/forum/view/help-support/rated-and-unrated-games-listing">chess.com forum</a>

<h4 dir="trl">created_at, last_move_at</h4>

<h5 dir="trl">טיפוס הנתונים</h5>

In [121]:
print('created_at dtype: ', orig_no_ids.created_at.dtype)
print('last_move_at dtype: ', orig_no_ids.last_move_at.dtype)

created_at dtype:  float64
last_move_at dtype:  float64


<h5 dir="trl">מספר ערכים שונים</h5>

In [122]:
print('created_at num of unique values: ', orig_no_ids.created_at.nunique())
print('last_move_at num of unique values: ', orig_no_ids.last_move_at.nunique())

created_at num of unique values:  12640
last_move_at num of unique values:  12670


<h5 dir="trl">ערכים לדוגמה</h5>

In [123]:
print('Some values from the created_at column: ', orig_no_ids.created_at.unique()[:5])
print('Some values from the last_move_at column: ', orig_no_ids.last_move_at.unique()[:5])

Some values from the created_at column:  [1.50421e+12 1.50413e+12 1.50411e+12 1.50403e+12 1.50424e+12]
Some values from the last_move_at column:  [1.50421e+12 1.50413e+12 1.50411e+12 1.50403e+12 1.50424e+12]


<h5 dir="trl">משמעות העמודות</h5>

<div dir="rtl">
<p>
אני משער שעמודות אלו מקודדות תאריך וזמן (כלומר את ה-
timestamp)
של אירועים כלשהם פר משחק. זה משאיר שתי שאלות פתוחות: מה האירועים, ומה הקידוד.
</p>
<p>
לגבי האירועים, אני מנחש ש-
last_move_at
זה התאריך והזמן שבו הסתיים המהלך האחרון במשחק.
קשה לי לומר מה משמעות
created_at.
האם הכוונה לתחילת המשחק או לאירוע אחר?
</p>
<p>
בין כך ובין כך, אני מעריך שהזמנים האבסולוטיים המקודדים בשתי העמודות הללו אינם חשובים,
אלא מה שחשוב זה ההפרש ביניהם, שנותן במדיוק או בערך את משך זמן המשחק.
אולם אם זה המצב, אז כמעט מחצית המשחקים התחילו והסתיימו באותו הזמן, כפי שמראה קטע הקוד הבא.
</p>
<div>

In [127]:
start_time_equals_end_time_row_num = sum(orig_no_ids.created_at == orig_no_ids.last_move_at)
print('Number of games that started and ended at the same time: ', start_time_equals_end_time_row_num)
print('Percentage of these games w.r.t the total number of games: ', (start_time_equals_end_time_row_num / orig_no_ids.shape[0]) * 100)

Number of games that started and ended at the same time:  8533
Percentage of these games w.r.t the total number of games:  44.64500601684717


<p dir="rtl">
נראה לי שעומדות בפני שתי ברירות:
למחוק מהטבלה את המשחקים שעבורם זמני תחילת וסוף המשחק זהים,
או למחוק מהטבלה את העמודות
created_at
ו-
last_move_at
מתוך הנחה שאני לא מבין את משמעות העמודות הללו,
או שהערכים המופיעים בהן "מקולקלים".
מאחר ומדובר במספר רב של שורות,
ואני גם לא יודע מה הקידוד ששימש ליצירת הערכים בעמודות אלה,
אני בוחר את האפשרות השניה.
</p>

In [128]:
orig_no_ids_no_times = orig_no_ids.drop(['created_at', 'last_move_at'], axis=1)

<h4 dir="trl">turns</h4>

<h4 dir="trl">victory_status</h4>

<h4 dir="trl">winner</h4>

<h4 dir="trl">increment_code</h4>

<h4 dir="trl">white_id</h4>

<h4 dir="trl">white_rating</h4>

<h4 dir="trl">black_id</h4>

<h4 dir="trl">black_rating</h4>

<h4 dir="trl">moves</h4>

<h4 dir="trl">opening_eco</h4>

<h4 dir="trl">opening_name</h4>

<h4 dir="trl">opening_ply</h4>

<h3 dir="trl">סיכום עמודות הטבלה</h3>

<div dir="rtl">
<table dir>
  <th>
    <td>תיאור קצר</td>
    <td>הערות</td>
  </th>
  <tr>
    <td>id</td>
    <td>מזהה של המשחק</td>
    <td>מסיבה לא ברורה מספר הערכים השונים בעמודה זו 19113 קטן ממספר השורות</td>
  </tr>
  <tr>
    <td>rated</td>
    <td>משחק שהוא rated הוא משחק שתוצאותיו משפיעות על דירוג השחקנים</td>
    <td />
  </tr>
  <tr>
    <td>created_at</td>
    <td />
    <td>לא ברורה לי משמעות עמודה זו: האם זה הזמן שבו התחיל המשחק, או הזמן שבו הוא "נוצר" והתחילו לשחק בו רק אח"כ</td>
  </tr>
  <tr>
    <td>last_move_at</td>
    <td>אני משער שעמודה זו מקודדת את התאריך והזמן שבו הסתיים המשחק</td>
    <td />
  </tr>
  <tr>
    <td>turns</td>
    <td>מספר התורות במשחק</td>
    <td>במשחק שחמט תור מורכב מזוג מהלכים: של הלבן ושל השחור</td>
  </tr>
  <tr>
    <td>victory_status</td>
    <td>4 אפשרויות: outoftime, resign, mate, draw</td>
    <td />
  </tr>
  <tr>
    <td>winner</td>
    <td>3 אפשרויות: black, white, draw</td>
    <td />
  </tr>
  <tr>
    <td>increment_code</td>
    <td>מחרוזת מהצורה: זמן משחק בדקות + תוספת זמן בשניות לכל מהלך, למשל 2+10 בשביל משחק שאורכו 2 דקות, עם 10 שניות הארכת זמן לכל מהלך</td>
    <td>כל פעם כשמגיע תורו של שחקן לשחק מתווספת הארכת הזמן לזמן שנותר לו לשחק</td>
  </tr>
  <tr>
    <td>white_id</td>
    <td>מזהה של השחקן הלבן</td>
    <td />
  </tr>
  <tr>
    <td>white_rating</td>
    <td>דירוג השחקן הלבן</td>
    <td>ככל שהדירוג גבוה יותר, כך השחקן טוב יותר</td>
  </tr>
  <tr>
    <td>black_id</td>
    <td>מזהה של השחקן השחור</td>
    <td />
  </tr>
  <tr>
    <td>black_rating</td>
    <td>דירוג השחקן השחור</td>
    <td>ככל שהדירוג גבוה יותר, כך השחקן טוב יותר</td>
  </tr>
  <tr>
    <td>moves</td>
    <td>תיאור שלם של המשחק בנוטציה הסטנדרטית של משחקי שחמט</td>
    <td />
  </tr>
  <tr>
    <td>opening_eco</td>
    <td>קוד המתאר את הפתיחה</td>
    <td />
  </tr>
  <tr>
    <td>opening_name</td>
    <td>שם הפתיחה, באנגלית</td>
    <td />
  </tr>
  <tr>
    <td>opening_ply</td>
    <td>מספר המהלכים בחלק הפתיחה</td>
    <td />
  </tr>
</table>
</div>